bindings/python/fdb/impl.py

```diff
@@ -1074,90 +1062,164 @@ def declare_asynchronous_transactions():
                 v = tr.get(key)
                 raise Return(v, tr.watch(key))
                 yield None
 
             @staticmethod
             @transactional
             @coroutine
-            def __database_set_and_watch(tr, key, value):
+            def __creator_set_and_watch(tr, key, value):
                 tr.set(key, value)
                 raise Return(tr.watch(key))
                 yield None
 
             @staticmethod
             @transactional
             @coroutine
-            def __database_clear_and_watch(tr, key):
+            def __creator_clear_and_watch(tr, key):
                 del tr[key]
                 raise Return(tr.watch(key))
                 yield None
 
             @staticmethod
             @transactional
             @coroutine
-            def __database_delitem(tr, key_or_slice):
+            def __creator_delitem(tr, key_or_slice):
                 del tr[key_or_slice]
                 raise Return()
                 yield None
 
             @staticmethod
             @transactional
             @coroutine
-            def __database_atomic_operation(tr, opcode, key, param):
+            def __creator_atomic_operation(tr, opcode, key, param):
                 tr._atomic_operation(opcode, key, param)
                 raise Return()
                 yield None
-        return Database
+        return TransactionCreator
+
+def process_tenant_name(name):
+    if isinstance(name, tuple):
+        return pack(name)
+    elif isinstance(name, bytes):
+        return name
+    else:
+        raise TypeError('Tenant name must be of type ' + bytes.__name__ + ' or of type ' + tuple.__name__)
+
+class Database(_TransactionCreator):
+    def __init__(self, dpointer):
+        self.dpointer = dpointer
+        self.options = _DatabaseOptions(self)
+
+    def __del__(self):
+        # print('Destroying database 0x%x' % self.dpointer)
+        self.capi.fdb_database_destroy(self.dpointer)
+
+    def _set_option(self, option, param, length):
+        self.capi.fdb_database_set_option(self.dpointer, option, param, length)
+
+    def open_tenant(self, name):
+        tname = process_tenant_name(name)
+        pointer = ctypes.c_void_p()
+        self.capi.fdb_database_open_tenant(self.dpointer, tname, len(tname), ctypes.byref(pointer))
+        return Tenant(pointer.value)
+
+    def create_transaction(self):
+        pointer = ctypes.c_void_p()
+        self.capi.fdb_database_create_transaction(self.dpointer, ctypes.byref(pointer))
+        return Transaction(pointer.value, self)
+
+    def allocate_tenant(self, name):
+        Database.__database_allocate_tenant(self, process_tenant_name(name), [])
+
+    def delete_tenant(self, name):
+        Database.__database_delete_tenant(self, process_tenant_name(name), [])
+
+    @staticmethod
+    @transactional
+    def __database_allocate_tenant(tr, name, existence_check_marker):
+        tr.options.set_special_key_space_enable_writes()
```

Comment:
I think I follow what's happening here, but would you mind adding a high-level description of the purpose of existence_check_marker? (I think the purpose is to only check if the tenant exists on the first attempt to commit)

Reply:
Yeah, I can add that